include/engine/datafacade/shared_datafacade.hpp

```diff
@@ -775,6 +810,35 @@ class SharedDataFacade final : public BaseDataFacade
         }
     }
 
+    // Returns the data source ids that were used to supply the edge
+    // weights.
+    virtual void
+    GetUncompressedReverseDatasources(const EdgeID id,
+                                      std::vector<uint8_t> &result_datasources) const override final
+    {
+        const unsigned begin = m_geometry_indices.at(id);
+        const unsigned end = m_geometry_indices.at(id + 1) - 1;
+
+        result_datasources.clear();
+        result_datasources.reserve(end - begin);
+
+        // If there was no datasource info, return an array of 0's.
+        if (m_datasource_list.empty())
+        {
+            for (unsigned i = 0; i < end - begin; ++i)
+            {
+                result_datasources.push_back(0);
+            }
+        }
+        else
+        {
+            std::for_each(
```

Comment:
@daniel-j-h what is your feeling here, could we make all of these accesses lazy copy views? Most times we will need all data, but for all the cases that only look at the first few items along a road, all this copying could be saved.